### ISSUE TITLE
Expose metrics as system sources

### ISSUE THREAD

**Issue by ruchirK** [MaterializeInc/materialize#1696]
Right now -- as far as I can tell -- one _needs_ to use grafana to inspect materialize metrics (such as # of records read via avro) 

it would be neat to add an option to have metrics also be a builtin source or piped to a file source -- and then we could inspect the metrics with materialize itself?

**Comment by rjnn** [comment]
I'd like to up the priority on this one. It's particularly useful for debugging crashed clusters, and I'd like to have this shipped before high priority crashes come up.

**Comment by benesch** [comment]
> It's particularly useful for debugging crashed clusters, and I'd like to have this shipped before high priority crashes come up.

The fact that you currently have to use Prometheus and Grafana in a production installation of Materialize is a feature, not a bug, in my opinion. Debugging a system with the system itself results in a world of pain. If someone installs a query that grinds the system to a halt—and there are a lot of such queries at the moment—Materialize's metric export and inspection pipeline will *also* grind to a halt, and you won't be able to debug.

I *do* think this feature is worthwhile, but I don't think the motivation should be to support anything remotely production related. This is Prometheus's bread and butter, and I think it's too early to start reinventing the monitoring wheel.

> But we shouldn't overthink this just yet. Making the logs be reasonable sized is something we can worry about when large logs are empirically shown to be a problem.

We don't need to think about it at all if the advice for now is to use Prometheus instead. Prometheus by default will purge things older than two weeks, for example.

**Comment by benesch** [comment]
It's also not clear to me that what's really desired is piping metrics to file sinks—that seems like a pretty roundabout way to get data that Materialize is itself generating into Materialize. If we had data persistence, then the feature would just be "expose metrics as a persistent source," and users could query them and create sinks based off on them just like they could for any other source.

**Comment by benesch** [comment]
Renaming to "expose metrics as system sources", as that is the more general feature, is accomplishable today with some elbow grease, and leaves the door open to installing file sinks on top of the sources by default—though as described above I really think we should be pushing people towards Prometheus/Grafana for the foreseeable future.

**Comment by cuongdo** [comment]
I agree with @benesch that most production monitoring should be done through Prometheus and Grafana. There's been a lot of work put into both, and past experience strongly suggests that it's difficult to diagnose software that's having issues using that same software.

There are two scenarios I can see where exposing metrics as system sources would be useful:

1. Initial exploration. When a user first downloads Materialize, it would be convenient to have access to  metrics without involving external processes.
2. Ad-hoc exploration of a running server. Combine this with SQL, and this unlocks aggregations and other features not easily achieved through Grafana.

**Comment by quodlibetor** [comment]
More detailed proposal that is related to this #5277

**Comment by antifuchs** [comment]
@quodlibetor and I met today to discuss this issue and how to shape a
fix for it, and I think we came up with a neat design that could
capture what we want & provide a good path forward for metrics in the
future. As it's really two designs, I'm not yet going to write it up
in proto-RFC format, but we can move each to the design doc stage
later.

# Design draft 1: Representing our Prometheus metrics internally

Introduce a prometheus `SOURCE`, which scrapes a prometheus endpoint
in an interval and imports each metric into tables not unlike this, keeping them up to :

``` sql
CREATE TABLE arbitrary_name_for_a_table_holding_metrics (
    metric TEXT,
    labels JSONB,
    value float8
);

CREATE TABLE arbitrary_name_for_a_table_holding_metrics_meta (
    metric TEXT,
    metric_type TEXT,
    help TEXT
);
```

(it's possible we might want to denormalize this, but as a prototype I
think this schema should do nicely; also, we might consider adding the
timestamp as of which the data is current; I'm leaving this out for
now to make it easier to think about the key structure).

Importing our own prometheus metrics (and maybe even others), we can
offer a very easy way from postgres-capable tooling to query and
visualize the operational metrics of a running system, without setting
up things like Prometheus ingestion or Grafana.

## Alternative approach we've considered

### Exposing an internal event source/sink via a channel

We could add a mechanism for providing log events to the dataflow
crate (or a crate that df and others depend on), where if a piece of
code now increments a prometheus counter, it would instead send an
event that gets captured by a listener on a channel, and increments
the prometheus cuonter, as well as queue the event for updating of
internal metrics views.

The major benefit is that we encapsulate our metrics tooling usage more
(making it easier to plug in more ways of recording metric values).

The downsides, however, are somewhat big:

* Things need to be structured such that you can get at the logger
  where you need to count the metric. Some parts where we currently do
  this make this somewhat difficult (e.g. the pgwire code, where the
  currently best way to count bytes sent is in encoding).

* Adding metrics gets more difficult. You not only have to figure out
  how go get your number counted the right way, you also have to
  figure out how to aggregate it properly in the dataflow world. We
  worry that this might mean that useful operational metrics don't get
  added as soon/as quickly as they'd be necessary, making it harder to
  run the system.

# Design draft 2: Getting system table data into Prometheus for alerting

Note: This is the other part of the puzzle ("every value we have on a log
table should be available as a prometheus metric"). We didn't spend as
much time speccing this out, but the design we have come up with does
sound plausible and desireable, so I'm writing it up.

Make a `SINK` that is a prometheus-formatted file (importable using
the node-exporter's [textfile
collector](https://github.com/prometheus/node_exporter#textfile-collector)).

So, given a view `customer_counts` that has columns "metric" and "value",
the following sink definition:

``` sql
CREATE SINK prometheus_from_customer_counts
FROM customer_counts
INTO PROMETHEUS '/run/prometheus_data/customer_counts.prom';
```

results in the file containing metrics named after the "metric"
column, value from the "value" column and with labels holding the
values of all the other columns.

It's plausible that we might send these to a pushgateway, but for now
files seems reasonable and easy; they might also fulfill the
requirement in #1573.

# Next steps
Of the two drafts, I think design 1 is just about ready to become
reality; then when it's proven out we can move towards exporting our
own metrics to prometheus as well via the second design.

**Comment by krishmanoh2** [comment]
On restart, would we loose history?

**Comment by cirego** [comment]
Given that these are two different design docs, would it make sense to make RFC PRs for each of these individually?

**Comment by antifuchs** [comment]
> On restart, would we loose history?

Yes - good call-out. The idea here is to expose the "current" values of the system, so until the scrape happens the first time, you wouldn't see the metric values. Steady-state you would get the last value (and no history), which should make this useful for debugging "in the moment" when an incident happens.

> Given that these are two different design docs, would it make sense to make RFC PRs for each of these individually?

Yep!

**Comment by ruchirK** [comment]
Can we close this issue now that there's a system table that captures prometheus metrics?

**Comment by antifuchs** [comment]
Oops, yes! I think I forgot to close this when I merged #6271.
